lib/ngx/semaphore.lua

```diff
@@ -0,0 +1,124 @@
+
+local ffi = require 'ffi'
+local base = require "resty.core.base"
+
+
+local FFI_OK = base.FFI_OK
+local FFI_ERROR = base.FFI_ERROR
+local FFI_DECLINED = base.FFI_DECLINED
+local ffi_new = ffi.new
+local ffi_str = ffi.string
+local ffi_gc = ffi.gc
+local C = ffi.C
+local type = type
+local error = error
+local tonumber = tonumber
+local getfenv = getfenv
+local get_string_buf = base.get_string_buf
+local get_size_ptr = base.get_size_ptr
+local setmetatable = setmetatable
+local co_yield = coroutine._yield
+local ERR_BUF_SIZE = 128
+
+
+local errmsg = base.get_errmsg_ptr()
+
+
+if not pcall(ffi.typeof,"ngx_http_lua_semaphore_t") then
+    ffi.cdef[[
+        struct ngx_http_lua_semaphore_s;
+        typedef struct ngx_http_lua_semaphore_s ngx_http_lua_semaphore_t;
+    ]]
+end
+
+ffi.cdef[[
+    int ngx_http_lua_ffi_semaphore_new(ngx_http_lua_semaphore_t **psem,
+        int n, char **errstr);
+    int ngx_http_lua_ffi_semaphore_post(ngx_http_lua_semaphore_t *sem,
+        int n, char **errstr);
+    int ngx_http_lua_ffi_semaphore_wait(ngx_http_request_t *r,
+        ngx_http_lua_semaphore_t *sem, int wait_ms, char *errstr, size_t *errlen);
+    void ngx_http_lua_ffi_semaphore_gc(ngx_http_lua_semaphore_t *sem);
+]]
+
+
+local _M = {}
+
+
+function _M.new(n)
+    local psem = ffi_new("ngx_http_lua_semaphore_t *[1]")
```

Comment:
This `psem` cdata should be lifted to the toplevel scope so that we don't have to allocate and free it on hot code paths. This is similar to the existing `errmsg` variable in the toplevel scope.